source/vir/src/interpreter.rs

```diff
@@ -0,0 +1,1391 @@
+                .iter()
+                .map(|b| {
+                    let name = b.name.clone();
+                    let a = eval_expr_internal(ctx, state, &b.a)?;
+                    Ok(Arc::new(BinderX { name, a }))
+                })
+                .collect();
+            let new_bnds = new_bnds?;
+            exp_new(Ctor(path.clone(), id.clone(), Arc::new(new_bnds)))
+        }
+        Interp(e) => match e {
+            InterpExp::FreeVar(_) => ok,
+            InterpExp::Seq(_) => ok,
+        },
+        // Ignored by the interpreter at present (i.e., treated as symbolic)
+        VarAt(..) | VarLoc(..) | Loc(..) | Old(..) | WithTriggers(..) => ok,
+    };
+    let res = r?;
+    state.depth -= 1;
+    state.log(format!("{}=> {:}", "\t".repeat(state.depth), &res));
+    if state.enable_simplified_cache {
+        state.simplified.insert(&res);
+    }
+    Ok(res)
+}
+
+/// We run the interpreter on a separate thread, so that we can give it a larger-than-default stack
+fn eval_expr_launch(
+    exp: Exp,
+    fun_ssts: SstMap,
+    rlimit: u32,
+    arch_size_min_bits: u32,
+    mode: ComputeMode,
+    log: Arc<Mutex<Option<File>>>,
+) -> Result<Exp, VirErr> {
+    let env = ScopeMap::new();
+    let cache = HashMap::new();
+    let mut file_log_opt = log.lock().unwrap();
+    let logging = file_log_opt.is_some();
+    let debug = logging;
+    let mut state = State {
+        depth: 0,
+        env,
+        iterations: 1,
+        log: file_log_opt.take(),
+        debug,
+        perf: logging,
+        cache,
+        enable_cache: true,
+        cache_hits: 0,
+        cache_misses: 0,
+        simplified: PtrSet::new(),
+        enable_simplified_cache: false,
+        ptr_hits: 0,
+        ptr_misses: 0,
+        fun_calls: HashMap::new(),
+    };
+    // Don't run for too long
+    let max_iterations =
+        if rlimit == 0 { std::u64::MAX } else { rlimit as u64 * RLIMIT_MULTIPLIER };
+    let ctx = Ctx { fun_ssts: &fun_ssts, max_iterations, arch_size_min_bits };
+    let res = eval_expr_internal(&ctx, &mut state, &exp)?;
+    display_perf_stats(&state);
+    if state.log.is_some() {
+        file_log_opt.replace(state.log.unwrap());
+    }
+    if let ExpX::Const(Constant::Bool(false)) = res.x {
+        err_str(&exp.span, "assert simplifies to false")
+    } else {
+        match mode {
+            // Send partial result to Z3
+            ComputeMode::Z3 => {
+                // Restore the free variables we hid during interpretation
+                let res = crate::sst_visitor::map_exp_visitor(&res, &mut |e| match &e.x {
+                    ExpX::Interp(InterpExp::FreeVar(v)) => {
+                        SpannedTyped::new(&e.span, &e.typ, ExpX::Var(v.clone()))
+                    }
+                    _ => e.clone(),
+                });
+                if debug {
+                    if exp.definitely_eq(&res) {
+                        println!();
```

Comment:
Can you use rustc's diagnostic output for this? (like we do for other notes/warnings)
This is enabled always when `--log-all` is enabled. If this is intended for internal debugging only, can you either print it to the .interp file or enable it separately (with a flag/env variable)

Reply:
If I correctly recall what we discussed, it may be helpful to emit this as a note (even when debugging is disabled) when you send a partially simplified expression to z3 and z3 fails to prove it.

Reply:
> Can you use rustc's diagnostic output for this?

Reply:
My take is: if it's output that the developer should see, it should go through `Diagnostics`, it should be associated with a span, and it shouldn't be conditional on whether `--log-*` is enabled (there is some machinery for VIR-generated info/warnings in `well_formed` but that needs to be extended if we need it elsewhere), if it's for debugging it should go to the log file.

If I understand this correctly (but it's very possible I don't), the "simplified to" message on failure is pretty similar to this debug info. Printing it only on failure seems like it would prevent noise when the proof goes through but still provide information to the user for debugging.

I believe the difference is that this debug info also mentions whether any simplification happened before sending the expression to z3. Ideally, I think we should print a warning (with a span) only when the computation does no simplification (if I understand correctly, this ends up being equivalent to a regular `assert`), so the user knows to either just use a regular `assert` or make changes so that computation is actually helpful.

Reply:
Just to summarize, it sounds like "Could not take advantage of compute to simplify expression before sending to z3" should always (when it applies) be printed using Diagnostics.  Then that leaves the "Was able to simplify before sending to Z3" message.  Right now, that's behind a `debug` flag, which I thought was primarily associated with the `--debug` flag.  It sounds like `--log-*` also enables it?  From the interpreter's perspective, if you enable `--log-interpreter`, then the simplified expression (if any) will be logged to the interpreter's log file, so printing it to the screen may be redundant.

Reply:
> Right now, that's behind a debug flag, which I thought was primarily associated with the --debug flag. It sounds like --log-* also enables it?

Unless I'm misunderstanding something, I don't think this output should be related to `--debug` which "Enable[s] debugging of proof failures".

`--log-all` does not enable `--debug`; your logic enables these `println`s when `--log-interpreter` (or `--log-all`) is enabled here https://github.com/secure-foundations/verus/blob/4db2241c4ed0c5022cd018dd69eee4ae4fcdc718/source/vir/src/interpreter.rs#L1329-L1349

I thought it looked like a bit of a roundabout way to do this but I guessed you were leaving this open for more configurability in the future. It looks like the `debug` flag in the interpreter's `State` is currently equivalent to `log.is_some()`. But it sounds like that wasn't the intent.

Reply:
Ah, I see what you mean!  Sorry, my cache on all of this was more completely flushed than I realized.  You're quite right that I apparently tied `debug` to logging.  I think `debug` initially toggled printing information to the terminal, but once we added the logging option, all of that goes to the log file.  Since `debug` doesn't seem to be doing much any more, I removed it, and now we only print a warning when the interpreter fails to do any simplification at all.

Reply:
Relevant commit: https://github.com/verus-lang/verus/pull/275/commits/6be031244a91d8d4c9f8d769fc30e1364feedca8